src/rgw/rgw_gc.cc

```diff
@@ -65,7 +65,64 @@ int RGWGC::tag_index(const string& tag)
   return rgw_shards_mod(XXH64(tag.c_str(), tag.size(), seed), max_objs);
 }
 
-int RGWGC::send_chain(cls_rgw_obj_chain& chain, const string& tag)
+int RGWGC::send_split_chain(const cls_rgw_obj_chain& chain, const std::string& tag)
+{
+  ldpp_dout(this, 20) << "RGWGC::send_split_chain - tag is: " << tag << dendl;
+
+  if (cct->_conf->rgw_max_chunk_size) {
+    cls_rgw_obj_chain broken_chain;
+    bool chain_sent = false;
+    size_t tag_size = tag.size(), cumulative_size = 0;
+    ldpp_dout(this, 20) << "RGWGC::send_split_chain - rgw_max_chunk_size is: " << cct->_conf->rgw_max_chunk_size << dendl;
+
+    for (auto it = chain.objs.begin(); it != chain.objs.end(); it++) {
+      auto size = it->estimate_encoded_size();
+      cumulative_size += size;
+      auto total_size = sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32) + sizeof(__u32) + cumulative_size; //adding overhead of encoding cls_rgw_obj_chain
+      total_size += sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32) + sizeof(__u32) + tag_size + (2 * sizeof(ceph_le32)); //adding overhead of encoding cls_rgw_gc_obj_info
+      total_size += sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32) + sizeof(ceph_le32); // adding overhead of encoding cls_rgw_gc_set_entry_op
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - cumulative_size is: " << cumulative_size << dendl;
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - total_size is: " << total_size << dendl;
+
+      if (total_size < cct->_conf->rgw_max_chunk_size) { //continue adding to chain
+        broken_chain.objs.emplace_back(*it);
+        chain_sent = false;
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - less than, add to broken chain and continue" << dendl;
+        continue;
+      }
+      if (total_size == cct->_conf->rgw_max_chunk_size) { //add to chain and send to gc
+        broken_chain.objs.emplace_back(*it);
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - equal to, add to broken chain and send chain" << dendl;
+        auto ret =  send_chain(broken_chain, tag);
+        if (ret < 0) {
+          ldpp_dout(this, 0) << "RGWGC::send_split_chain - send chain returned error: " << ret << dendl;
+          return ret;
+        }
+        chain_sent = true;
+      }
+      if (total_size > cct->_conf->rgw_max_chunk_size) { //dont add to chain, and send to gc
+        --it;
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - more than, dont add to broken chain and send chain" << dendl;
+        auto ret = send_chain(broken_chain, tag);
+        if (ret < 0) {
+          ldpp_dout(this, 0) << "RGWGC::send_split_chain - send chain returned error: " << ret << dendl;
+          return ret;
+        }
+        chain_sent = true;
+      }
+      broken_chain.objs.clear();
+      cumulative_size = 0;
+    }
+    if (!chain_sent) { //when the chain is smaller than rgw_max_chunk_size
+      send_chain(broken_chain, tag);
```

Comment:
need to check for error, or return it directly

Reply:
done.